### ISSUE TITLE
Have concurrency option should increase the speed at which secrets are fetch.

### ISSUE THREAD

**Issue by samof76** [external-secrets/external-secrets#330]
**Describe the solution you'd like**
Having a lot secrets to synced at a give epoch, looks like the fetch is sequentional. Is there an arg or option to make these get calls parallel? 

**What is the added value?**
The update of secrets will be much faster if they happened in parallel.

**Give us examples of the outcome**
some option like 

```
--concurrency=n
```
to your code will distribute the secrets to be fectch across some `n` goroutines.

**Observations (Constraints, Context, etc):**

EKS cluster running on 1.18
Running ExS 0.2.14

**Comment by samof76** [comment]
We have about 700+ ExternalSecrets each with a refresh interval of `15m` and if I see the logs I see this...
```
2021-08-19T13:25:26.911Z    DEBUG    controllers.ExternalSecret    reconciled secret  ... 
2021-08-19T13:25:26.920Z    DEBUG    provider.aws    using region    ...                                            
2021-08-19T13:25:26.920Z    DEBUG    provider.aws    assuming role    ...         
2021-08-19T13:25:26.920Z    INFO    provider.aws.secretsmanager    fetching secret value   ...
2021-08-19T13:25:30.133Z    DEBUG    controllers.ExternalSecret    reconciled secret    ...
2021-08-19T13:25:30.141Z    DEBUG    provider.aws    using region    ...                                         
2021-08-19T13:25:30.141Z    DEBUG    provider.aws    assuming role    ...         
2021-08-19T13:25:30.141Z    INFO    provider.aws.secretsmanager    ...            
2021-08-19T13:25:33.362Z    DEBUG    controllers.ExternalSecret    reconciled secret    ...
2021-08-19T13:25:33.371Z    DEBUG    provider.aws    using region    ...                                         
2021-08-19T13:25:33.371Z    DEBUG    provider.aws    assuming role    ...         
2021-08-19T13:25:33.371Z    INFO    provider.aws.secretsmanager    fetching secret value    ...
2021-08-19T13:25:36.594Z    DEBUG    controllers.ExternalSecret    reconciled secret    ...
2021-08-19T13:25:36.602Z    DEBUG    provider.aws    using region    ...                                         
2021-08-19T13:25:36.602Z    DEBUG    provider.aws    assuming role    ...         
2021-08-19T13:25:36.602Z    INFO    provider.aws.secretsmanager    fetching secret value    ...
2021-08-19T13:25:39.796Z    DEBUG    controllers.ExternalSecret    reconciled secret    ...
2021-08-19T13:25:39.807Z    DEBUG    provider.aws    using region    ...                                         
2021-08-19T13:25:39.807Z    DEBUG    provider.aws    assuming role    ...         
2021-08-19T13:25:39.807Z    INFO    provider.aws.secretsmanager    fetching secret value    ...
2021-08-19T13:25:42.979Z    DEBUG    controllers.ExternalSecret    reconciled secret    ...
2021-08-19T13:25:42.997Z    DEBUG    provider.aws    using region    ...                                         
2021-08-19T13:25:42.997Z    DEBUG    provider.aws    assuming role    ...         
2021-08-19T13:25:42.997Z    INFO    provider.aws.secretsmanager    fetching secret value    ...
2021-08-19T13:25:46.219Z    DEBUG    controllers.ExternalSecret    reconciled secret    ...
2021-08-19T13:25:46.226Z    DEBUG    provider.aws    using region    ...                                         
2021-08-19T13:25:46.226Z    DEBUG    provider.aws    assuming role    ...         
2021-08-19T13:25:46.226Z    INFO    provider.aws.secretsmanager    fetching secret value    ...

```

And many times if I update the a secrets it takes more than 15 mins to get updated, close to 30 minutes.

**Comment by jack-evans** [comment]
I guess there is 2 ways you could go in terms of concurrency, and it would be good to distinguish what you exactly mean, do you mean concurrency in handling multiple External secret reconciles in parallel or do you mean handling multiple gathering of secrets (defined within the external secrets CR) in parallel? 

The former would definitely be possible however the code should be checked for any issues accessing object concurrently

**Comment by knelasevero** [comment]
Yeah, what @jack-evans said :) 

Also, do you need secrets to be updated this quickly? Or is the refresh interval small just because you want to be sure that when you change a secret in a provider they will be updated in the cluster in less than 15 min? If you don't expect changes every 15 min, and you just want changes to go when you have new values, you can force updates with annotations (probably only for the secrets that were changed, if you want). Like `kubectl annotate es my-es force-sync=$(date +%s) --overwrite`

Another thing, are all those 700+ ESs from the same "group of services"? Are they all in the same namespace? The other option would be to use the cotrollerClass approach and have multiple operators running, and having them handle stores with the right controller key set.

Just exploring some options. 

But anyways some concurrency options are interesting and we will have a look into it :)

**Comment by samof76** [comment]
ExS - ExternalSecret
ASM - AWS SecretsManager
> do you mean concurrency in handling multiple External secret reconciles in parallel
Yes
> do you mean handling multiple gathering of secrets (defined within the external secrets CR) in parallel
I think is well handled already. BTW we have one-one mapping between an ExS and an ASM secret.

> Also, do you need secrets to be updated this quickly?
Not really. But we would definitely want a mechanism for an ExS CR to update a K8s `Secret` as soon as is done on the ASM. But your method seems to involve a manual mechanism, that is why we could want to keep the `refreshInterval` low.

> Another thing, are all those 700+ ESs from the same "group of services"?
Yes they all belong to the same namespace or "group of services".

**Comment by knelasevero** [comment]
Please have a look over this discussion: https://github.com/external-secrets/external-secrets/issues/129

You can make it automatic without manual step from the Aws side

**Comment by jack-evans** [comment]
@samof76 I have a branch [here](https://github.com/jack-evans/external-secrets/tree/concurrency) but have not tested for any concurrent object access issues (however a brief look through the code I couldn't see any).

I would be grateful if you could test it with your 700 ExternalSecrets on a test cluster to validate that it is working better. Note you may have to increase cpu and mem limits as concurrent reconciles will make them the bottleneck.

With helm chart you should be able to set `extraArgs` like the following:

```yaml
extraArgs:
- concurrent: 4
```

**Comment by samof76** [comment]
@jack-evans will do this tomorrow and let you know.

**Comment by samof76** [comment]
> Please have a look over this discussion: #129
> 
> Presumably you can make it automatic without manual step from the Aws side, or with something else like a Cron checking changes and only annotating changes.

This is really excellent bit of conversation that somehow needs to find its way into the doc. May be having an FAQ in the docs may help.

**Comment by samof76** [comment]
May tomorrow or friday will have update on this.

**Comment by jack-evans** [comment]
@samof76 had a chance to test my changes yet?

**Comment by jack-evans** [comment]
> I wonder if it's worth implementing some kind of mechanism on the level of the SecretStore. 
> 
> If enabled, it should pull the list of secrets with their timestamps, and keep it in memory. This way we can avoid doing unneccessary trips for individual secrets if they have not been changed. 
> 
> 
> 
> (Note: I have no idea at the moment if this op would be supported by all providers or not)

I think this would be useful too, or something where we cache secrets retrieved with a small TTL (5mins say) so that any secrets retrieved multiple times in a short period can be utilised without spamming the backing secretStore?

**Comment by jack-evans** [comment]
@knelasevero i have the code in a branch but was waiting for @samof76 to test, happy to update my branch and put a PR in though if we think there won't be any side effects?

**Comment by jsarkar** [comment]
@samof76 did you able to test concurrent reconciles?
@jack-evans I am volunteering test this with Hashi vault backend. Will keep you posted

**Comment by gusfcarvalho** [comment]
I do believe this issue can now be addressed in #735, considering reconcile loop testing is indeed working. Please keep posting any update on it. I'll leave this one open for a little while now, and then we can switch to the new issue if you won't mind :)
